### PR TITLE
<includes> and <excludes> filtering feature for extraDirectories

### LINKED ISSUE #2564: Allow to add single file to a container (with setting rights)
**Environment**:

- *Jib version:* 2.4.0
- *Build tool:* maven:3.6.3-amazoncorretto-11
- *OS:* Amazon Linux 2

**Description of the issue**:

It would be useful if it was possible to add single file to a container (with setting rights). I am currently migrating from one system to the other and for some time I need to keep things compatible between both of them. Currently it is not possible (without workaround) to move the following Dockerfile command to jib-maven-plugin (it's possible only for directories):

COPY ./ssh/id_rsa /home/myuser/.ssh/id_rsa

**Steps to reproduce**:

    <extraDirectories>
        <paths>
            <path>
                <from>./ssh/id_rsa</from>
                <into>/home/myuser/.ssh/id_rsa</into>
             </path>
        </paths>
    </extraDirectories>

Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.4.0:build (default-cli) on project config: /home/builder/src/config/ssh/id_rsa is not a directory -> [Help 1]

### PR DESCRIPTION
**Pull request by chanseokoh**
On top of #3171. Review it first.

Fixes #2564 (and #3164).

The idea is to allow glob patterns that works similar to how [`maven-resources-plugin`](https://maven.apache.org/plugins/maven-resources-plugin/examples/include-exclude.html) works. This makes adding a single file easy as well.

```xml
         <extraDirectories>
            <paths>
              <path>
                <from>src/main/resources/static</from>
                <into>/app/static</into>
                <includes>foo.txt</includes> <!-- only foo.txt will be copied -->
              </path>
              <path>
                <from>src/main/resources/pictures</from>
                <into>/app/pictures</into>
                <!-- only copy these, but exclude some. -->
                <includes>**/*.bmp,**/*.gif</includes>
                <excludes>
                  <exclude>**/*test*.*</exclude>
                  <exclude>**/*.thumbnail</exclude>
                </excludes>
              </path>
            </paths>
          </extraDirectories>
```

In a follow-up PR, I'll add integration tests for Maven and Gradle and update CHANGELOG.

### CODE REVIEW

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java [hunk 1]
```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
...
               }
-            });
+            }
+
+            // Add with default permissions
+            builder.addEntry(localPath, pathOnContainer, modificationTime);
+          } else {
+            // Add with explicit permissions
+            builder.addEntry(localPath, pathOnContainer, permissions, modificationTime);
+          }
```
Comment (mpeddada1) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:104]:
```suggestion
          } 
            // Add with explicit permissions
            builder.addEntry(localPath, pathOnContainer, permissions, modificationTime);
```

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java [hunk 2]
```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
...
+              targetDirectory.resolve(sourceDirectory.relativize(localPath));
+          Instant modificationTime = modificationTimeProvider.get(localPath, pathOnContainer);
+          FilePermissions permissions = extraDirectoryPermissions.get(pathOnContainer.toString());
+          if (permissions == null) {
+            // Check for matching globs
+            Path containerPath = Paths.get(pathOnContainer.toString());
+            for (Map.Entry<PathMatcher, FilePermissions> entry : pathMatchers.entrySet()) {
+              if (entry.getKey().matches(containerPath)) {
+                builder.addEntry(localPath, pathOnContainer, entry.getValue(), modificationTime);
+                return;
```
Comment (mpeddada1) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:95]: there seem to be too many levels of nesting here, can we put this in a helper method?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java [hunk 3]
```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
...
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .forEach(pathMatcher -> walker.filter(path -> !pathMatcher.matches(path)));
+    includes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .map(pathMatcher -> (Predicate<Path>) (path -> pathMatcher.matches(path)))
+        .reduce((matches1, matches2) -> matches1.or(matches2))
+        .ifPresent(walker::filter);
+    walker.walk(
+        localPath -> {
```
Comment (mpeddada1) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:84]: Hm this method is starting to get a little long, can we have line 83 - 105 in a separate method?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java [hunk 4]
```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
...
+    excludes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .forEach(pathMatcher -> walker.filter(path -> !pathMatcher.matches(path)));
+    includes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .map(pathMatcher -> (Predicate<Path>) (path -> pathMatcher.matches(path)))
+        .reduce((matches1, matches2) -> matches1.or(matches2))
+        .ifPresent(walker::filter);
```
Comment (mpeddada1) [inline jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:82]: I'm probably missing something here, why is the `reduce` step needed here?

Comment (chanseokoh) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:82]: So before `reduce`, we have a list of `Predicate`s where each predicate tests if a given path matches its `<include>` glob pattern (i.e., testing if the file should be included).

Then `reduce` converts a list to a single `Predicate` that is contributed by "or"ing all the predicates. So, this predicate tests if a given path matches any of the `<include>` patterns. And `walker::filter()` accepts a single predicate. Note calling `walker::filter` repeatedly for each `Predicate` before `reduce()` makes them connected with "and", not "or".

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java (hunk 3 above)
Comment (chanseokoh) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:88]: The closure makes use of a lot of external variables, and mechanically refactoring out the closure results in a long list of arguments.
```java
  private static void addLayerEntry(
      Path sourceDirectory,
      AbsoluteUnixPath targetDirectory,
      Map<String, FilePermissions> extraDirectoryPermissions,
      ModificationTimeProvider modificationTimeProvider,
      Map<PathMatcher, FilePermissions> permissionPathMatchers,
      FileEntriesLayer.Builder layerBuilder,
      Path localPath) {
```
I realized most of the complication is due to determining permissions and calling `addEntry` based on that. So I refactored out the permissions part.

I also added a few more code comments in this method.

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java (hunk 1 above)
Comment (chanseokoh) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java]: Doesn't matter in the refactored method, but to be clear, `else` is needed unless you add `return` in the "true" branch.

File: jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java [hunk 5]
```diff
@@ -92,17 +80,91 @@ public void testExtraDirectoryLayerConfiguration() throws URISyntaxException, IO
...
+    Path extraFilesDirectory = Paths.get(Resources.getResource("core/layer").toURI());
+    FileEntriesLayer layerConfiguration =
+        JavaContainerBuilderHelper.extraDirectoryLayerConfiguration(
+            extraFilesDirectory,
+            AbsoluteUnixPath.get("/"),
+            Arrays.asList("**/*"),
+            Arrays.asList("**/*"),
+            Collections.emptyMap(),
+            (ignored1, ignored2) -> Instant.EPOCH);
+    assertThat(layerConfiguration.getEntries()).isEmpty();
```
Comment (mpeddada1) [inline jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java:150]: ah so excludes takes precedence?

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java (hunk 3 above)
Comment (mpeddada1) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:88]: Looks good. Putting the permissions logic in a helper helped a lot

File: jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java (hunk 4 above)
Comment (mpeddada1) [reply jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java:85]: Makes sense. Thanks for the explanation!

**Review by mpeddada1 — APPROVED** [review]

File: jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java (hunk 5 above)
Comment (chanseokoh) [reply jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java:150]: Yeah, if something is excluded, then that's absolute and final. I just need to double-check the behavior with the maven-resource-plugin.

Comment (chanseokoh) [reply jib-plugins-common/src/test/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelperTest.java:150]: Just verified that the behavior is the same as the Maven resources plugin.

**Pull request merged**